### ISSUE TITLE
Feature: developing a context handler

### ISSUE THREAD

**Issue by Cube707** [magmax/python-readchar#92]
Due to the way readchar currently handles terminal setup, we cannot control terminal behaviour outside of the two readchar  functions. This leads to issues # and #

We could solve this by externalising terminal setup and teardown and have it run at the beginning and end, ensuring consistent behaviour in between. Under python this calls for a `Context Manager`

Requirements would be:

- [ ] setting up the terminal for readchars needs, this means:
    - no echoing or other handeling of typed keys. They need to be waiting untill readchar calls read the next time
    - no interruptions, readchar will read  `CTRL+C` etc. As normal keystrokesand handle them
    - whatever else is needet to ensure readchar works consistently
- [ ] giving a way to check if a key is waiting to be read
- [ ] reading chars and keys as with the current behavior
- [ ] ensuring consistent behaviour above both platform's (even if that means limiting one)
- [ ] decide on good, intuitive names
- [ ] document the new feature

Here is an example of how it could work:

```python
With ReadChar() as read:
    If read.key_iswaiting():
        print(read.key())

    # user can do other stuff here in parallel #
    sleep(0.5)
``` 

If this is working an properly set up, I could replace the existing functions and they could be simply wrappers for this.

**Comment by Cube707** [comment]
@petereon i opened a new issue to discuss this. Feel free to drop more questions

**Comment by petereon** [comment]
Perfect, thank you very much for an exhaustive description. I wonder if it doesn't sound like a good opportunity to utilize `Queues` and reactive paradigm.

That would give us a very good way to check if there is a character waiting for processing with a FIFO `Queue`'s `get` method. With asyncio, it should also deal with the back pressure, if someone typed faster than it is able to process mentioned in #73 . For some reference, consider [this article](https://pierrepaci.medium.com/reactive-stream-in-pure-python-a01b41df934) and specifically this [code snippet](https://gist.github.com/PPACI/4aa3704f42b6a01baaaf4116c02a43c8#file-reactive-queue-py). Does that sound like something that could fit into the current architecture?

**Comment by Cube707** [comment]
While this would defnetly be a reasonable approach where we to rewrite all of he input gathering yourselfs.

But lukily the operating system already handles most of the heavy lifting. We just need to configure it to match our needs. So don't overcomplicate it.

**Comment by petereon** [comment]
Yeah, that sounds like a pretty sane approach. I am just unsure how to handle the none-blocking read then. Sounds like there is either going to have to be an async event-loop or separate process/thread in that case.

**Comment by Cube707** [comment]
No, the terminal is already a queue and you can query it for content. On windows its super simple as there is a function for it. `msvcrt.kbhit()` returns true if there are keypresses waiting to be read.

On linux use this https://gist.github.com/michelbl/efda48b19d3e587685e3441a74457024#file-kbhit-py-L103-L111

**Comment by Cube707** [comment]
Also check out this development version that I made but didn't use

https://github.com/Cube707/python-readchar/blob/c3081ec51581ead9b4db2f60a8f1567ed809d06a/readchar/posix_read.py

https://github.com/Cube707/python-readchar/blob/c3081ec51581ead9b4db2f60a8f1567ed809d06a/readchar/win_read.py

**Comment by petereon** [comment]
Great, thank you for the information and apologies for how very uninformed I am about `tty`. I'll take a deep dive in this stuff in coming days 😃

**Comment by Cube707** [comment]
No worry. The reason I never used the development version is that I didn't know enough and thought it would work just like that. Turned out its more complicated 😂

**Comment by Cube707** [comment]
Also keep in mind that on the windows side a lot of stuff doesn't work, so this will mostlikley be the limiting factor. So don't spend to much time (unless you want to 😆)

**Comment by petereon** [comment]
I will have a proper read into this. Good opportunity to learn how OS's handle this stuff :smile:. It might take a bit of time.

**Comment by Cube707** [comment]
take your time. there is no rush. 

I will also have more time in the coming weeks, as my thesis is due tomorrow and no longer eat up my schedule

**Comment by petereon** [comment]
Good luck with your thesis, mate. So glad I am done with that stuff already
😃

**Comment by Cube707** [comment]
I spend some time on this and prepeard  the function definitions withour content. See: https://github.com/magmax/python-readchar/commit/a5cecc3c7276cde73a37dba4422d93551464394f

**Comment by petereon** [comment]
Thanks, let me have a look.

**Comment by Cube707** [comment]
did some more work, check https://github.com/magmax/python-readchar/tree/contexmanager

**Comment by petereon** [comment]
That's very cool, also I can see you are already done by the time I got to it. Impressive. Apologies for my inactivity, I was really hard pressed for time with work related stuff.

**Comment by Cube707** [comment]
I would still be happy about feedback. Especially about the intuitivenes of names and the general usability

**Comment by petereon** [comment]
I have played around with it and had one issue in terms of usability: When there are multiple keys waiting at the same time, once the first is read it behaves as if no more keys were waiting. To replicate this, try the following code, press multiple keys while sleeping and wait for while loop to start running. Only printed the first pressed keys for me. After I pressed another as the while loop was running it dumped all the other ones.
```python
import time
from readchar import ReadChar

with ReadChar() as stream:
    time.sleep(5)
    while True:
        if stream.key_waiting:
            char = stream.key()
            print(char)
```


In terms of naming conventions, I wondered if this isn't more of a `CharStream` or `CharPipe` rather than `ReadChar`.

**Comment by Cube707** [comment]
I will look into the issue. I assumr you tested on Linux, not Windows?

For the name, my initial idea was something like ` KeyReader`, but I chose this as it is how other libary tend to name their context managers that have a normal function a counterpart. I also figured that it would be among the first thing people guess, alog the lines of "what was the CM from `readchar`? Probably just a capital `R`". But I might be totally off here...

**Comment by petereon** [comment]
I tested on macOS. But that's still POSIX I guess.

As for the naming, that does sound like a good way to approach it. Would be my first guess also.
